src/main/scala/pipedsl/passes/ExnTranslationPass.scala

```diff
@@ -0,0 +1,131 @@
+package pipedsl.passes
+
+import pipedsl.common.Syntax._
+import pipedsl.passes.Passes.{ModulePass, ProgPass}
+import pipedsl.typechecker.BaseTypeChecker.replaceNamedType
+
+object ExnTranslationPass extends ModulePass[ModuleDef] with ProgPass[Prog]{
+  private var exnArgMap = Map[Id, Id]()
+
+  override def run(m: ModuleDef): ModuleDef =
+  {
+    val fixed_commit = m.commit_blk match {
+      case Some(c) => c
+      case _ => CEmpty()
+    }
+
+    val fixed_except = m.except_blk match {
+      case ExceptFull(_, c) => c
+      case ExceptEmpty() => CEmpty()
+    }
+
+    val new_m = addExnVars(m)
+    new_m.name.typ = m.name.typ
+    print(new_m.body)
+    val modified_exnblk = m.except_blk.map(convertExnArgsId)
+    createNewStg(new_m.copy(body = new_m.body, commit_blk = new_m.commit_blk, except_blk = modified_exnblk))
+  }
+
+  override def run(p: Prog): Prog = p.copy(moddefs = p.moddefs.map(m => run(m)))
+
+  def addExnVars(m: ModuleDef): ModuleDef =
+  {
+    val fixed_except = m.except_blk match {
+      case ExceptFull(args, c) =>
+        var arg_count = 0
+        args.foreach(arg => {
+          exnArgMap = exnArgMap + (arg -> Id("_exnArg_"+arg_count.toString()))
```

Comment:
This `Id` needs a type  annotation (same as `arg`)

Reply:
Also it might be better to make it a map from `Id -> Expr` and then store `EVar`s so that you don't need to re-set the types of each of the times you create a new `EVar` from these `Id`s.